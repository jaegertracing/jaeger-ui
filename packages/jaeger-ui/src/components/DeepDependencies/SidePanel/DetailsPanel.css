/*
Copyright (c) 2020 The Jaeger Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.Ddg--DetailsPanel {
  background-color: #fafafa;
  border-left: solid 1px #ddd;
  box-shadow: -3px 0 3px rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  height: 100%;
}

.Ddg--DetailsPanel--SvcOpHeader {
  border-bottom: solid 1px #ddd;
  box-shadow: 0px 3px 3px rgba(0, 0, 0, 0.1);
  font-size: 1.5em;
  text-align: right;
  padding: 0.1em 0.3em;
}

.Ddg--DetailsPanel--DecorationHeader {
  font-size: 1.3em;
  text-align: center;
  padding: 0.3em;
}

.Ddg--DetailsPanel--DecorationHeader > span {
  border-bottom: solid 1px #bbb;
  box-shadow: 0px 3px 3px -3px rgba(0, 0, 0, 0.1);
}

.Ddg--DetailsPanel--DetailLink {
  padding-left: 0.2em;
}

.Ddg--DetailsPanel--errorMsg {
  color: #e58c33;
}

.Ddg--DetailsPanel--DetailsCard {
  background-color: #ffffff;
  border: solid 2px rgba(0, 0, 0, 0.3);
  margin: 0.5em;
  overflow: hidden;
}

.Ddg--DetailsPanel--DetailsCard.is-error {
  color: #e58c33;
}

.Ddg--DetailsPanel--LoadingIndicator {
  display: block;
  margin: auto;
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  line-height: 0;
}

.Ddg--DetailsPanel--LoadingWrapper {
  flex-grow: 1;
  position: relative;
}

.Ddg--DetailsPanel--PercentCircleWrapper {
  margin: 0 auto;
  max-width: 20vh;
  padding: 0 3%;
}
