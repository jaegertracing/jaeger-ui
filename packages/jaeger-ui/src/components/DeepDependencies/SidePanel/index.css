/*
Copyright (c) 2020 The Jaeger Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.Ddg--SidePanel {
  background-color: #f6f6f6;
  border-left: solid 1px #ddd;
  box-shadow: -3px 0 3px rgba(0, 0, 0, 0.1);
  display: flex;
}

.Ddg--SidePanel--Btns,
.Ddg--SidePanel--DecorationBtns {
  display: flex;
  flex-direction: column;
}

.Ddg--SidePanel--Btns {
  justify-content: space-between;
}

.Ddg--SidePanel--Btns button {
  cursor: pointer;
}

.Ddg--SidePanel--closeBtn,
.Ddg--SidePanel--infoBtn {
  background: transparent;
  border: none;
  margin: 5px 0px;
}

.Ddg--SidePanel--closeBtn > svg,
.Ddg--SidePanel--infoBtn > svg {
  height: 80%;
  width: 80%;
}

.Ddg--SidePanel--closeBtn:focus > svg,
.Ddg--SidePanel--infoBtn:focus > svg {
  filter: drop-shadow(0 0 2px #006c99);
}

.Ddg--SidePanel--closeBtn:focus,
.Ddg--SidePanel--infoBtn:focus {
  outline: none;
}

.Ddg--SidePanel--closeBtn.is-hidden {
  visibility: hidden;
  pointer-events: none;
}

.Ddg--SidePanel--DecorationBtns {
  flex-grow: 1;
  justify-content: center;
}

.Ddg--SidePanel--decorationBtn {
  border-radius: 100%;
  margin: 3px;
  padding-bottom: 25%;
  padding-top: 25%;
}

.Ddg--SidePanel--decorationBtn.is-selected {
  box-shadow: 0px 0px 4px 2px #006c99;
}

.Ddg--SidePanel--decorationBtn:focus {
  box-shadow: inset 0 0 5px 2px #00b4ff;
  outline: none;
}

.Ddg--SidePanel--decorationBtn.is-selected:focus {
  border: none;
  margin: 4px;
  box-shadow: inset 0 0 5px 2px #00b4ff, 0px 0px 5px 3px #006c99;
}
