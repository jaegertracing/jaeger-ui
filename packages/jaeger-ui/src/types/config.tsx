// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TNil } from '.';

export type ConfigMenuItem = {
  label: string;
  url: string;
  anchorTarget?: '_self' | '_blank' | '_parent' | '_top';
};

export type ConfigMenuGroup = {
  label: string;
  items: ConfigMenuItem[];
};

export type TScript = {
  text: string;
  type: 'inline';
};

export type LinkPatternsConfig = {
  type: 'process' | 'tags' | 'logs' | 'traces';
  key?: string;
  url: string;
  text: string;
};

export type Config = {
  archiveEnabled?: boolean;
  deepDependencies?: { menuEnabled?: boolean };
  dependencies?: { dagMaxServicesLen?: number; menuEnabled?: boolean };
  menu: (ConfigMenuGroup | ConfigMenuItem)[];
  search?: { maxLookback: { label: string; value: string }; maxLimit: number };
  scripts?: TScript[];
  topTagPrefixes?: string[];
  tracking?: {
    cookieToDimension?: {
      cookie: string;
      dimension: string;
    }[];
    gaID: string | TNil;
    trackErrors: boolean | TNil;
  };
  linkPatterns?: LinkPatternsConfig;
};
