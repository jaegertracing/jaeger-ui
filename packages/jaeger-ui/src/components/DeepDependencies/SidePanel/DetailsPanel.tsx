// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Tooltip } from 'antd';
import _get from 'lodash/get';
import { connect } from 'react-redux';

import BreakableText from '../../common/BreakableText';
import DetailsCard from '../../common/DetailsCard';
import LoadingIndicator from '../../common/LoadingIndicator';
import NewWindowIcon from '../../common/NewWindowIcon';
import VerticalResizer from '../../common/VerticalResizer';
import JaegerAPI from '../../../api/jaeger';
import extractDecorationFromState, { TDecorationFromState } from '../../../model/path-agnostic-decorations';
import stringSupplant from '../../../utils/stringSupplant';

import { TPathAgnosticDecorationSchema } from '../../../model/path-agnostic-decorations/types';
import { TColumnDefs, TDetails } from '../../common/DetailsCard/types';

import './DetailsPanel.css';

type TProps = TDecorationFromState & {
  decorationSchema: TPathAgnosticDecorationSchema;
  service: string;
  operation?: string | string[] | null;
};

type TState = {
  columnDefs?: TColumnDefs;
  details?: TDetails;
  detailsErred?: boolean;
  detailsLoading?: boolean;
  width?: number;
};

export class UnconnectedDetailsPanel extends React.PureComponent<TProps, TState> {
  state: TState = {};

  componentDidMount() {
    this.fetchDetails();
  }

  componentDidUpdate(prevProps: TProps) {
    if (
      prevProps.operation !== this.props.operation ||
      prevProps.service !== this.props.service ||
      prevProps.decorationSchema !== this.props.decorationSchema
    ) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        details: undefined,
        detailsErred: false,
        detailsLoading: false,
      });
      this.fetchDetails();
    }
  }

  fetchDetails() {
    const {
      decorationSchema: {
        detailUrl,
        detailPath,
        detailColumnDefPath,
        opDetailUrl,
        opDetailPath,
        opDetailColumnDefPath,
      },
      operation: _op,
      service,
    } = this.props;

    const operation = _op && !Array.isArray(_op) ? _op : undefined;

    let fetchUrl: string | undefined;
    let getDetailPath: string | undefined;
    let getDefPath: string | undefined;
    if (opDetailUrl && opDetailPath && operation) {
      fetchUrl = stringSupplant(opDetailUrl, { service, operation });
      getDetailPath = stringSupplant(opDetailPath, { service, operation });
      getDefPath = opDetailColumnDefPath && stringSupplant(opDetailColumnDefPath, { service, operation });
    } else if (detailUrl && detailPath) {
      fetchUrl = stringSupplant(detailUrl, { service });
      getDetailPath = stringSupplant(detailPath, { service });
      getDefPath = detailColumnDefPath && stringSupplant(detailColumnDefPath, { service });
    }

    if (!fetchUrl || !getDetailPath) return;

    this.setState({ detailsLoading: true });

    JaegerAPI.fetchDecoration(fetchUrl)
      .then((res: unknown) => {
        let detailsErred = false;
        let details = _get(res, getDetailPath as string);
        if (details === undefined) {
          details = `\`${getDetailPath}\` not found in response`;
          detailsErred = true;
        }
        const columnDefs: TColumnDefs = getDefPath ? _get(res, getDefPath, []) : [];

        this.setState({ columnDefs, details, detailsErred, detailsLoading: false });
      })
      .catch((err: Error) => {
        this.setState({
          details: `Unable to fetch decoration: ${err.message || err}`,
          detailsErred: true,
          detailsLoading: false,
        });
      });
  }

  onResize = (width: number) => {
    this.setState({ width });
  };

  render() {
    const { decorationProgressbar, decorationSchema, decorationValue, operation: _op, service } = this.props;
    const { detailLink } = decorationSchema;
    const { width = 0.3 } = this.state;
    const operation = _op && !Array.isArray(_op) ? _op : undefined;
    return (
      <div className="Ddg--DetailsPanel" style={{ width: `${width * 100}vw` }}>
        <div className="Ddg--DetailsPanel--SvcOpHeader">
          <BreakableText text={service} />
          {operation && <BreakableText text={`::${operation}`} />}
        </div>
        <div className="Ddg--DetailsPanel--DecorationHeader">
          <span>{stringSupplant(decorationSchema.name, { service, operation })}</span>
          {detailLink && (
            <Tooltip arrowPointAtCenter title="More Info">
              <a
                className="Ddg--DetailsPanel--DetailLink"
                href={stringSupplant(detailLink, { service, operation })}
                target="_blank"
                rel="noreferrer noopener"
              >
                <NewWindowIcon />
              </a>
            </Tooltip>
          )}
        </div>
        {decorationProgressbar ? (
          <div className="Ddg--DetailsPanel--PercentCircleWrapper">{decorationProgressbar}</div>
        ) : (
          <span className="Ddg--DetailsPanel--errorMsg">{decorationValue}</span>
        )}
        {this.state.detailsLoading && (
          <div className="Ddg--DetailsPanel--LoadingWrapper">
            <LoadingIndicator className="Ddg--DetailsPanel--LoadingIndicator" />
          </div>
        )}
        {this.state.details && (
          <DetailsCard
            className={`Ddg--DetailsPanel--DetailsCard ${this.state.detailsErred ? 'is-error' : ''}`}
            columnDefs={this.state.columnDefs}
            details={this.state.details}
            header="Details"
          />
        )}
        <VerticalResizer max={0.8} min={0.1} onChange={this.onResize} position={width} rightSide />
      </div>
    );
  }
}

export default connect(extractDecorationFromState)(UnconnectedDetailsPanel);
