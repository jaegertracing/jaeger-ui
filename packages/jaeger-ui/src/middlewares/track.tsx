// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Action } from 'redux-actions';
import { Dispatch, Store } from 'react-redux';

import { middlewareHooks as searchHooks } from '../components/SearchTracePage/SearchForm.track';
import { middlewareHooks as timelineHooks } from '../components/TracePage/TraceTimelineViewer/duck.track';
import { isWaEnabled } from '../utils/tracking';
import { ReduxState } from '../types';

type TMiddlewareFn = (store: Store<ReduxState>, action: Action<any>) => void;

const middlewareHooks: { [actionType: string]: TMiddlewareFn } = { ...timelineHooks, ...searchHooks };

function trackingMiddleware(store: Store<ReduxState>) {
  return function inner(next: Dispatch<ReduxState>) {
    return function core(action: any) {
      const { type } = action;
      if (typeof middlewareHooks[type] === 'function') {
        middlewareHooks[type](store, action);
      }
      return next(action);
    };
  };
}

export default isWaEnabled ? trackingMiddleware : undefined;
