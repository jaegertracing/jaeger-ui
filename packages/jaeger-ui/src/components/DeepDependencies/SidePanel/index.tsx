// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Modal, Table } from 'antd';
import { MdExitToApp, MdInfoOutline } from 'react-icons/md';

import { TDdgVertex } from '../../../model/ddg/types';
import { TPathAgnosticDecorationSchema } from '../../../model/path-agnostic-decorations/types';
import { getConfigValue } from '../../../utils/config/get-config';
import DetailsPanel from './DetailsPanel';
import * as track from './index.track';

import './index.css';

type TProps = {
  clearSelected: () => void;
  selectDecoration: (decoration?: string) => void;
  selectedDecoration?: string;
  selectedVertex?: TDdgVertex;
};

export default class SidePanel extends React.PureComponent<TProps> {
  decorations: TPathAgnosticDecorationSchema[] | undefined;

  constructor(props: TProps) {
    super(props);

    const { selectedDecoration, selectedVertex } = props;
    if (selectedDecoration) track.trackDecorationSelected(selectedDecoration);
    if (selectedVertex) track.trackDecorationViewDetails(selectedVertex);

    this.decorations = getConfigValue('pathAgnosticDecorations');
  }

  componentDidUpdate(prevProps: TProps) {
    if (prevProps.selectedVertex !== this.props.selectedVertex) {
      track.trackDecorationViewDetails(this.props.selectedVertex);
    }
  }

  clearSelected = () => {
    track.trackDecorationViewDetails();
    this.props.clearSelected();
  };

  selectDecoration = (decoration?: string) => {
    track.trackDecorationSelected(decoration);
    this.props.selectDecoration(decoration);
  };

  openInfoModal = () => {
    Modal.info({
      content: (
        <Table
          columns={[
            {
              dataIndex: 'acronym',
              key: 'acronym',
              title: 'Acronym',
            },
            {
              dataIndex: 'name',
              key: 'name',
              title: 'Name',
            },
          ]}
          dataSource={this.decorations}
          rowKey={(schema: TPathAgnosticDecorationSchema) => schema.id}
        />
      ),
      maskClosable: true,
      title: 'Decoration Options',
      width: '60vw',
    });
  };

  render() {
    if (!this.decorations) return null;

    const { selectedDecoration, selectedVertex } = this.props;

    const selectedSchema = this.decorations.find(({ id }) => id === selectedDecoration);

    return (
      <div className="Ddg--SidePanel">
        <div className="Ddg--SidePanel--Btns">
          <button
            className={`Ddg--SidePanel--closeBtn ${selectedVertex && selectedSchema ? '' : 'is-hidden'}`}
            type="button"
            onClick={this.clearSelected}
          >
            <MdExitToApp />
          </button>
          <div className="Ddg--SidePanel--DecorationBtns">
            {this.decorations.map(({ acronym, id }) => (
              <button
                key={id}
                className={`Ddg--SidePanel--decorationBtn ${id === selectedDecoration ? 'is-selected' : ''}`}
                type="button"
                onClick={() => this.selectDecoration(id)}
              >
                {acronym}
              </button>
            ))}
            <button
              key="clearBtn"
              className="Ddg--SidePanel--decorationBtn"
              type="button"
              onClick={() => this.selectDecoration()}
            >
              Clear
            </button>
          </div>
          <button className="Ddg--SidePanel--infoBtn" onClick={this.openInfoModal} type="button">
            <MdInfoOutline />
          </button>
        </div>
        <div className={`Ddg--SidePanel--Details ${selectedVertex && selectedSchema ? '.is-expanded' : ''}`}>
          {selectedVertex && selectedSchema && (
            <DetailsPanel
              decorationSchema={selectedSchema}
              operation={selectedVertex.operation}
              service={selectedVertex.service}
            />
          )}
        </div>
      </div>
    );
  }
}
