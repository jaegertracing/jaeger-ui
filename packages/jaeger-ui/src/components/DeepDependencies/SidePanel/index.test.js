// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Modal } from 'antd';

import SidePanel from '.';
import * as track from './index.track';
import * as getConfig from '../../../utils/config/get-config';

describe('<SidePanel>', () => {
  let getConfigValueSpy;
  const testAcronym = 'TA';
  const testID = 'test ID';
  const mockConfig = [
    {
      id: 'first',
      acronym: '1st',
      name: 'First Decoration',
    },
    {
      id: testID,
      acronym: testAcronym,
      name: 'Decoration to test interactions',
    },
    {
      id: 'last',
      acronym: 'LO',
      name: 'The last one',
    },
  ];
  const testVertex = { service: 'svc', operation: 'op' };

  let trackDecorationSelectedSpy;
  let trackDecorationViewDetailsSpy;

  beforeAll(() => {
    getConfigValueSpy = jest.spyOn(getConfig, 'getConfigValue').mockReturnValue(mockConfig);
    trackDecorationSelectedSpy = jest.spyOn(track, 'trackDecorationSelected');
    trackDecorationViewDetailsSpy = jest.spyOn(track, 'trackDecorationViewDetails');
  });

  beforeEach(() => {
    trackDecorationSelectedSpy.mockReset();
    trackDecorationViewDetailsSpy.mockReset();
  });

  describe('constructor', () => {
    it('inits decorations', () => {
      const wrapper = shallow(<SidePanel />);
      expect(wrapper.instance().decorations).toBe(mockConfig);
    });

    it('tracks initial selection', () => {
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(0);

      shallow(<SidePanel selectedDecoration={testID} />);
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(1);
      expect(trackDecorationSelectedSpy).toHaveBeenLastCalledWith(testID);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(0);

      shallow(<SidePanel selectedVertex={testVertex} />);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith(testVertex);
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('render', () => {
    it('renders null if there are no decorations', () => {
      getConfigValueSpy.mockReturnValueOnce(undefined);
      const wrapper = shallow(<SidePanel />);
      expect(wrapper.getElement()).toBe(null);
    });

    it('renders config decorations with clear button', () => {
      const wrapper = shallow(<SidePanel />);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders selected decoration', () => {
      const wrapper = shallow(<SidePanel selectedDecoration={testID} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('ignores selectedVertex without selected decoration', () => {
      const wrapper = shallow(<SidePanel selectedVertex={testVertex} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders sidePanel and closeBtn when vertex and decoration are both selected', () => {
      const wrapper = shallow(<SidePanel selectedDecoration={testID} selectedVertex={testVertex} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentDidUpdate', () => {
    it('tracks change in vertex from absent to present', () => {
      const wrapper = shallow(<SidePanel />);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(0);

      wrapper.setProps({ selectedVertex: testVertex });
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith(testVertex);
    });

    it('tracks change in vertex from present to absent', () => {
      const wrapper = shallow(<SidePanel selectedVertex={testVertex} />);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);

      wrapper.setProps({ selectedVertex: undefined });
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(2);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith(undefined);
    });

    it('tracks change in vertex between different vertexes', () => {
      const wrapper = shallow(<SidePanel selectedVertex={testVertex} />);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith(testVertex);

      const newVertex = { ...testVertex };
      wrapper.setProps({ selectedVertex: newVertex });
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(2);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith(newVertex);
    });

    it('does not track unchanged vertex', () => {
      const wrapper = shallow(<SidePanel selectedVertex={testVertex} />);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith(testVertex);

      wrapper.setProps({ selectedDecoration: testID });
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('clearSelected', () => {
    it('clears selected and tracks clearing', () => {
      const clearSelected = jest.fn();
      const wrapper = shallow(<SidePanel selectedVertex={testVertex} clearSelected={clearSelected} />);
      expect(clearSelected).toHaveBeenCalledTimes(0);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(1);

      wrapper
        .find('button')
        .at(0)
        .simulate('click');
      expect(clearSelected).toHaveBeenCalledTimes(1);
      expect(trackDecorationViewDetailsSpy).toHaveBeenCalledTimes(2);
      expect(trackDecorationViewDetailsSpy).toHaveBeenLastCalledWith();
    });
  });

  describe('selectDecoration', () => {
    const selectDecoration = jest.fn();

    beforeEach(() => {
      selectDecoration.mockReset();
    });

    it('selects decoration and tracks selection', () => {
      const wrapper = shallow(<SidePanel selectDecoration={selectDecoration} />);
      expect(selectDecoration).toHaveBeenCalledTimes(0);
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(0);

      wrapper.find(`button[children="${testAcronym}"]`).simulate('click');
      expect(selectDecoration).toHaveBeenCalledTimes(1);
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(1);
      expect(trackDecorationSelectedSpy).toHaveBeenLastCalledWith(testID);
    });

    it('clears decoration and tracks clear', () => {
      const wrapper = shallow(<SidePanel selectDecoration={selectDecoration} selectedDecoration={testID} />);
      expect(selectDecoration).toHaveBeenCalledTimes(0);
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(1);

      wrapper
        .find('.Ddg--SidePanel--DecorationBtns > button')
        .last()
        .simulate('click');
      expect(selectDecoration).toHaveBeenCalledTimes(1);
      expect(trackDecorationSelectedSpy).toHaveBeenCalledTimes(2);
      expect(trackDecorationSelectedSpy).toHaveBeenLastCalledWith(undefined);
    });
  });

  describe('info button ', () => {
    let modalInfoSpy;

    beforeAll(() => {
      modalInfoSpy = jest.spyOn(Modal, 'info');
    });

    it('opens info modal', () => {
      const wrapper = shallow(<SidePanel />);
      expect(modalInfoSpy).toHaveBeenCalledTimes(0);

      wrapper
        .find('button')
        .last()
        .simulate('click');
      expect(modalInfoSpy).toHaveBeenCalledTimes(1);
      expect(modalInfoSpy.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
