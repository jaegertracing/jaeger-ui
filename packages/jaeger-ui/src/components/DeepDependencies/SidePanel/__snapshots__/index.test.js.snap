// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidePanel> info button  opens info modal 1`] = `
Object {
  "content": <Table
    columns={
      Array [
        Object {
          "dataIndex": "acronym",
          "key": "acronym",
          "title": "Acronym",
        },
        Object {
          "dataIndex": "name",
          "key": "name",
          "title": "Name",
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "acronym": "1st",
          "id": "first",
          "name": "First Decoration",
        },
        Object {
          "acronym": "TA",
          "id": "test ID",
          "name": "Decoration to test interactions",
        },
        Object {
          "acronym": "LO",
          "id": "last",
          "name": "The last one",
        },
      ]
    }
    rowKey={[Function]}
  />,
  "maskClosable": true,
  "title": "Decoration Options",
  "width": "60vw",
}
`;

exports[`<SidePanel> render ignores selectedVertex without selected decoration 1`] = `
<div
  className="Ddg--SidePanel"
>
  <div
    className="Ddg--SidePanel--Btns"
  >
    <button
      className="Ddg--SidePanel--closeBtn is-hidden"
      onClick={[Function]}
      type="button"
    >
      <IoExitOutline />
    </button>
    <div
      className="Ddg--SidePanel--DecorationBtns"
    >
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="first"
        onClick={[Function]}
        type="button"
      >
        1st
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="test ID"
        onClick={[Function]}
        type="button"
      >
        TA
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="last"
        onClick={[Function]}
        type="button"
      >
        LO
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn"
        key="clearBtn"
        onClick={[Function]}
        type="button"
      >
        Clear
      </button>
    </div>
    <button
      className="Ddg--SidePanel--infoBtn"
      onClick={[Function]}
      type="button"
    >
      <IoInformationCircleOutline />
    </button>
  </div>
  <div
    className="Ddg--SidePanel--Details "
  />
</div>
`;

exports[`<SidePanel> render renders config decorations with clear button 1`] = `
<div
  className="Ddg--SidePanel"
>
  <div
    className="Ddg--SidePanel--Btns"
  >
    <button
      className="Ddg--SidePanel--closeBtn is-hidden"
      onClick={[Function]}
      type="button"
    >
      <IoExitOutline />
    </button>
    <div
      className="Ddg--SidePanel--DecorationBtns"
    >
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="first"
        onClick={[Function]}
        type="button"
      >
        1st
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="test ID"
        onClick={[Function]}
        type="button"
      >
        TA
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="last"
        onClick={[Function]}
        type="button"
      >
        LO
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn"
        key="clearBtn"
        onClick={[Function]}
        type="button"
      >
        Clear
      </button>
    </div>
    <button
      className="Ddg--SidePanel--infoBtn"
      onClick={[Function]}
      type="button"
    >
      <IoInformationCircleOutline />
    </button>
  </div>
  <div
    className="Ddg--SidePanel--Details "
  />
</div>
`;

exports[`<SidePanel> render renders selected decoration 1`] = `
<div
  className="Ddg--SidePanel"
>
  <div
    className="Ddg--SidePanel--Btns"
  >
    <button
      className="Ddg--SidePanel--closeBtn is-hidden"
      onClick={[Function]}
      type="button"
    >
      <IoExitOutline />
    </button>
    <div
      className="Ddg--SidePanel--DecorationBtns"
    >
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="first"
        onClick={[Function]}
        type="button"
      >
        1st
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn is-selected"
        key="test ID"
        onClick={[Function]}
        type="button"
      >
        TA
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="last"
        onClick={[Function]}
        type="button"
      >
        LO
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn"
        key="clearBtn"
        onClick={[Function]}
        type="button"
      >
        Clear
      </button>
    </div>
    <button
      className="Ddg--SidePanel--infoBtn"
      onClick={[Function]}
      type="button"
    >
      <IoInformationCircleOutline />
    </button>
  </div>
  <div
    className="Ddg--SidePanel--Details "
  />
</div>
`;

exports[`<SidePanel> render renders sidePanel and closeBtn when vertex and decoration are both selected 1`] = `
<div
  className="Ddg--SidePanel"
>
  <div
    className="Ddg--SidePanel--Btns"
  >
    <button
      className="Ddg--SidePanel--closeBtn "
      onClick={[Function]}
      type="button"
    >
      <IoExitOutline />
    </button>
    <div
      className="Ddg--SidePanel--DecorationBtns"
    >
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="first"
        onClick={[Function]}
        type="button"
      >
        1st
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn is-selected"
        key="test ID"
        onClick={[Function]}
        type="button"
      >
        TA
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn "
        key="last"
        onClick={[Function]}
        type="button"
      >
        LO
      </button>
      <button
        className="Ddg--SidePanel--decorationBtn"
        key="clearBtn"
        onClick={[Function]}
        type="button"
      >
        Clear
      </button>
    </div>
    <button
      className="Ddg--SidePanel--infoBtn"
      onClick={[Function]}
      type="button"
    >
      <IoInformationCircleOutline />
    </button>
  </div>
  <div
    className="Ddg--SidePanel--Details .is-expanded"
  >
    <Connect(UnconnectedDetailsPanel)
      decorationSchema={
        Object {
          "acronym": "TA",
          "id": "test ID",
          "name": "Decoration to test interactions",
        }
      }
      operation="op"
      service="svc"
    />
  </div>
</div>
`;
